### PR TITLE
[Tyra2] Physics directory

### PR DESCRIPTION
**Pull request by Wellinator**
Implements Ray;
Implements min and max in BBox class;

### CODE REVIEW

File: engine/inc/renderer/3d/bbox/bbox.hpp [hunk 1]
```diff
@@ -41,9 +41,21 @@ class BBox : public CoreBBox {
...
+   * @param res_min Vec4 to store the min result
+   * @param res_max Vec4 to store the min result
+   * @brief Calc and stores the min and max points of box at a single loop
+   * */
+  void calcMinMax(Vec4& res_min, Vec4& res_max);
+
  protected:
   float _height, _depth, _width;
-  Vec4 _centerVector;
+  Vec4 _centerVector, _min;
```
Comment (h4570) [inline engine/inc/renderer/3d/bbox/bbox.hpp]: `_min` is probably not necessary?

File: engine/inc/physics/ray.hpp [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+namespace Tyra {
+
+/**
+ * Class for raycasting.
+ */
+class Ray {
+ public:
+  /**
+   * Constructor create a ampty ray
+   *
```
Comment (h4570) [inline engine/inc/physics/ray.hpp]: Unnecessary newline

File: engine/inc/physics/ray.hpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+   */
+  Ray();
+
+  /**
+   * Constructor
+   * @param origin The origin of the ray
+   * @param direction The direction of the ray (normalized Vec4)
+   *
```
Comment (h4570) [inline engine/inc/physics/ray.hpp]: Unnecessary newline

File: engine/inc/physics/ray.hpp [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+  Vec4 origin;
+  Vec4 direction;
+
+  // Methods
+
+  /**
+   * Set origin and direction
+   */
+  void set(Vec4 origin, Vec4 direction);
```
Comment (h4570) [inline engine/inc/physics/ray.hpp]: `const Vec4&` should be set as parameter in all of `setX()` functions

File: engine/inc/physics/ray.hpp [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+   * Set direction only
+   */
+  void setDirection(Vec4 origin);
+  void setDirection(const float& t_x, const float& t_y, const float& t_z);
+
+  /**
+   * Return Vec4 that is a given distance along this Ray
+   * @param t - the distance along the Ray to retrieve a position for.
+   */
+  Vec4 at(float t);
```
Comment (h4570) [inline engine/inc/physics/ray.hpp]: `const float&`

File: engine/inc/physics/ray.hpp [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+  /**
+   * Return the distance from the Vec4 point to the origin
+   */
+  float distanceToPoint(Vec4 point);
```
Comment (h4570) [inline engine/inc/physics/ray.hpp]: `const Vec4&`

File: engine/inc/physics/ray.hpp [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+  float distanceTo(const Vec4& v) const;
+
+  /**
+   * Returns Vec4 of intersection position
+   * @param minCorner - pointer to min corner position. (bottom left)
+   * @param maxCorner - pointer to max corner position. (top right)
+   * @param distance - the box to intersect with.
+   *
+   */
+  u8 intersectBox(Vec4* minCorner, Vec4* maxCorner, float& distance);
```
Comment (h4570) [inline engine/inc/physics/ray.hpp]: Why `Vec4*`, instead of `const Vec4&`?

File: engine/inc/physics/ray.hpp [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+  /**
+   * Returns distance from origin to intersected mesh
+   */
+  inline const Vec4 getPosition() { return _position; };
```
Comment (h4570) [inline engine/inc/physics/ray.hpp]: `inline const Vec4&`

File: engine/inc/renderer/3d/bbox/bbox.hpp [hunk 9]
```diff
@@ -41,9 +41,21 @@ class BBox : public CoreBBox {
   /** @returns the bottom face (nearer on y-axis) */
   const BBoxFace& getBottomFace() { return _bottomFace; }
 
+  /** @returns the lower (x, y, z) boundary of the box. */
+  const Vec4 min();
```
Comment (h4570) [inline engine/inc/renderer/3d/bbox/bbox.hpp]: Why `const`? Lets give programmer full choice what he want to do with it.
(the same for `max()`)

File: engine/inc/renderer/3d/bbox/bbox.hpp [hunk 10]
```diff
@@ -41,9 +41,21 @@ class BBox : public CoreBBox {
...
+  /** @returns the upper (x, y, z) boundary of the box. */
+  const Vec4 max();
+  /**
+   * @returns void;
+   * @param res_min Vec4 to store the min result
+   * @param res_max Vec4 to store the min result
+   * @brief Calc and stores the min and max points of box at a single loop
+   * */
+  void calcMinMax(Vec4& res_min, Vec4& res_max);
```
Comment (h4570) [inline engine/inc/renderer/3d/bbox/bbox.hpp]: I suggest renaming to `getMinMax(..)` and changing `Vec4&` to `Vec4*`

File: engine/inc/renderer/3d/bbox/bbox.hpp [hunk 11]
```diff
@@ -41,9 +41,21 @@ class BBox : public CoreBBox {
   /** @returns the bottom face (nearer on y-axis) */
   const BBoxFace& getBottomFace() { return _bottomFace; }
 
+  /** @returns the lower (x, y, z) boundary of the box. */
+  const Vec4 min();
+  /** @returns the upper (x, y, z) boundary of the box. */
+  const Vec4 max();
+  /**
+   * @returns void;
```
Comment (h4570) [inline engine/inc/renderer/3d/bbox/bbox.hpp]: This line is unnecessary

**Review by h4570 — CHANGES_REQUESTED** [review]
Only cosmetic changes

File: engine/inc/physics/ray.hpp (hunk 7 above)
Comment (h4570) [reply engine/inc/physics/ray.hpp]: Please try use following rule (as parameter):
`const Vec4&` -> we will not change your `Vec4` in this function
`Vec4*` -> we will change your `Vec4` in this function for sure!

File: engine/inc/physics/ray.hpp (hunk 8 above)
Comment (Wellinator) [reply engine/inc/physics/ray.hpp]: This function is not needed anymore, I'll remove it;

File: engine/inc/renderer/3d/bbox/bbox.hpp (hunk 1 above)
Comment (Wellinator) [reply engine/inc/renderer/3d/bbox/bbox.hpp]: Yeah, it's not uset. I used an local variable intead... gonna remove it.

File: engine/inc/renderer/3d/bbox/bbox.hpp (hunk 10 above)
Comment (Wellinator) [reply engine/inc/renderer/3d/bbox/bbox.hpp]: I used getMinMax but changed to calcMinMax, lol;
I'll set the old name, sounds better...

File: engine/inc/renderer/3d/bbox/bbox.hpp (hunk 9 above)
Comment (Wellinator) [reply engine/inc/renderer/3d/bbox/bbox.hpp]: Right!

**Comment by Wellinator** [comment]
Fixed. c:

**Review by h4570 — APPROVED** [review]

**Pull request merged**